File: src/signals/whale.ts
/**
 * Whale Alert signal fetcher.
 *
 * Fetches transactions >$3M from the Whale Alert API and classifies them
 * as accumulation or distribution signals by destination type.
 *
 * Key insight: not all large transfers are equal.
 *   - Exchange inflow  = likely selling (bearish for that token)
 *   - Exchange outflow = likely accumulation into cold storage (bullish)
 *   - Unknown wallet   = ambiguous, weighted lower
 *   - Miner wallet     = selling pressure (bearish)
 *
 * We maintain a rolling 2h net flow window per symbol.
 * The whale tracker strategy (strategies/whale-tracker.ts) reads from this.
 */

import { env } from '../config.js';
import { log } from '../storage/database.js';
import { onWhaleTransfer } from '../strategies/whale-tracker.js';

// ─── Types ────────────────────────────────────────────────────────────────────

type WalletType = 'exchange' | 'unknown_wallet' | 'known_fund' | 'miner';

interface WATransaction {
  blockchain: string;
  symbol: string;
  id: string;
  transaction_type: string;
  hash: string;
  from: { address: string; owner?: string; owner_type: string };
  to: { address: string; owner?: string; owner_type: string };
  timestamp: number;
  amount: number;
  amount_usd: number;
}

interface WAResponse {
  result: string;
  count: number;
  transactions: WATransaction[];
}

// ─── Owner type normalizer ────────────────────────────────────────────────────

function toWalletType(ownerType: string): WalletType {
  switch (ownerType.toLowerCase()) {
    case 'exchange':   return 'exchange';
    case 'fund':
    case 'custodian':  return 'known_fund';
    case 'miner':      return 'miner';
    default:           return 'unknown_wallet';
  }
}

// ─── Fetch loop ────────────────────────────────────────────────────────────────

const MIN_USD = 3_000_000;
const seenTxIds = new Set<string>();
let lastCursor = 0;
let lastFetchAt = 0;
const POLL_INTERVAL_MS = 120_000; // 2 min

export async function pollWhaleAlerts(symbols: string[]): Promise<void> {
  if (!env.whaleAlertApiKey) return;
  if (Date.now() - lastFetchAt < POLL_INTERVAL_MS) return;

  const since = lastCursor || Math.floor((Date.now() - 7_200_000) / 1000); // default: last 2h
  const url = `https://api.whale-alert.io/v1/transactions?api_key=${env.whaleAlertApiKey}&min_value=${MIN_USD}&start=${since}&limit=100`;

  try {
    const res = await fetch(url, { signal: AbortSignal.timeout(8_000) });
    if (!res.ok) {
      log('warn', `Whale Alert fetch failed: ${res.status}`);
      return;
    }
    const data = await res.json() as WAResponse;
    if (data.result !== 'success') return;

    let newCursor = lastCursor;

    for (const tx of data.transactions) {
      if (seenTxIds.has(tx.id)) continue;
      seenTxIds.add(tx.id);

      const sym = tx.symbol.toUpperCase();
      if (!symbols.includes(sym)) continue;

      newCursor = Math.max(newCursor, tx.timestamp);

      onWhaleTransfer({
        symbol: sym,
        amountUsd: tx.amount_usd,
        fromType: toWalletType(tx.from.owner_type),
        toType: toWalletType(tx.to.owner_type),
        knownWallet: tx.from.owner ?? tx.to.owner,
        ts: tx.timestamp * 1000,
      });

      log('info', `Whale: $${(tx.amount_usd / 1e6).toFixed(0)}M ${sym} ${tx.from.owner_type} → ${tx.to.owner_type}`, {
        symbol: sym,
        data: { amountUsd: tx.amount_usd, from: tx.from.owner_type, to: tx.to.owner_type },
      });
    }

    lastCursor = newCursor;
    lastFetchAt = Date.now();

    // Prevent unbounded growth
    if (seenTxIds.size > 5_000) {
      const arr = Array.from(seenTxIds);
      seenTxIds.clear();
      for (const id of arr.slice(-2_000)) seenTxIds.add(id);
    }
  } catch (err) {
    log('warn', `Whale Alert network error: ${String(err)}`);
  }
}

File: src/signals/whale.py
"""Whale Alert large transaction fetcher."""

import time

import requests

from src.config import env
from src.signals._circuit_breaker import CircuitBreaker
from src.storage.database import log

MIN_USD = 3_000_000
_seen_tx_ids: set[str] = set()
_last_cursor: int = 0
_last_fetch_at: float = 0
_POLL_INTERVAL_MS = 120_000
_breaker = CircuitBreaker("whale")


def _to_wallet_type(owner_type: str) -> str:
    ot = owner_type.lower()
    if ot == "exchange":
        return "exchange"
    if ot in ("fund", "custodian"):
        return "known_fund"
    if ot == "miner":
        return "miner"
    return "unknown_wallet"


def poll_whale_alerts(symbols: list[str]) -> None:
    global _last_cursor, _last_fetch_at
    if not env.whale_alert_api_key:
        return
    now = time.time() * 1000
    if now - _last_fetch_at < _POLL_INTERVAL_MS:
        return

    # Staleness warning
    if _last_fetch_at > 0 and now - _last_fetch_at > 2 * _POLL_INTERVAL_MS:
        log("warn", f"Whale alert data is stale (last fetch {(now - _last_fetch_at) / 60_000:.0f}m ago)")

    if not _breaker.can_call():
        log("warn", "Whale alert circuit breaker OPEN — skipping poll")
        return

    since = _last_cursor or int((now - 7_200_000) / 1000)
    url = (
        f"https://api.whale-alert.io/v1/transactions"
        f"?api_key={env.whale_alert_api_key}&min_value={MIN_USD}&start={since}&limit=100"
    )

    try:
        res = requests.get(url, timeout=8)
        if res.status_code != 200:
            log("warn", f"Whale Alert fetch failed: {res.status_code}")
            _breaker.record_failure()
            return
        data = res.json()
        if data.get("result") != "success":
            _breaker.record_failure()
            return

        _breaker.record_success()

        # Lazy import to avoid circular dependency
        from src.strategies.whale_tracker import on_whale_transfer

        new_cursor = _last_cursor
        for tx in data.get("transactions", []):
            tx_id = tx["id"]
            if tx_id in _seen_tx_ids:
                continue
            _seen_tx_ids.add(tx_id)

            sym = tx["symbol"].upper()
            if sym not in symbols:
                continue

            new_cursor = max(new_cursor, tx["timestamp"])
            on_whale_transfer({
                "symbol": sym,
                "amount_usd": tx["amount_usd"],
                "from_type": _to_wallet_type(tx["from"]["owner_type"]),
                "to_type": _to_wallet_type(tx["to"]["owner_type"]),
                "known_wallet": tx["from"].get("owner") or tx["to"].get("owner"),
                "ts": tx["timestamp"] * 1000,
            })

            log("info",
                f"Whale: ${tx['amount_usd'] / 1e6:.0f}M {sym} {tx['from']['owner_type']} -> {tx['to']['owner_type']}",
                symbol=sym,
                data={"amount_usd": tx["amount_usd"], "from": tx["from"]["owner_type"], "to": tx["to"]["owner_type"]})

        _last_cursor = new_cursor
        _last_fetch_at = now

        if len(_seen_tx_ids) > 5000:
            keep = list(_seen_tx_ids)[-2000:]
            _seen_tx_ids.clear()
            _seen_tx_ids.update(keep)

    except Exception as err:
        log("warn", f"Whale Alert network error: {err}")
        _breaker.record_failure()
